feat: Implement `waitFor`

eps1lon opened the pull request:
Closes https://github.com/testing-library/web-testing-library/issues/3
[API preview](https://deploy-preview-1204--testing-library.netlify.app/docs/web-testing-library/api#waitfor)
Docs: https://github.com/testing-library/testing-library-docs/pull/1204

eps1lon committed:
feat: Rough sketch for API (includes untested implementation)

eps1lon commented on src/waitFor.tsx:
```diff
@@ -0,0 +1,172 @@
+      // eslint-disable-next-line no-unmodified-loop-condition, @typescript-eslint/no-unnecessary-condition
+      while (!finished && !signal?.aborted) {
+        clock.advanceTimersByTime(interval)
+
+        // It's really important that checkCallback is run *before* we flush
+        // in-flight promises. To be honest, I'm not sure why, and I can't quite
+        // think of a way to reproduce the problem in a test, but I spent
+        // an entire day banging my head against a wall on this.
+        checkCallback()
```
Don't need this anymore. It's implicitly called inside  `clock.advanceTimersByTime(interval)` due to the above `const intervalId = setInterval(checkCallback, interval)`
```suggestion
```

timdeschryver commented on src/waitFor.tsx:
```diff
@@ -0,0 +1,172 @@
+// This is so the stack trace the developer sees is one that's
```
Does this need to be a `*.tsx` file?

eps1lon committed:
Can be in .ts

eps1lon committed:
Test reference impl

eps1lon committed:
Full test suite

eps1lon committed:
Ensure aborted signal works

eps1lon committed:
Gate AbortController tests

We also test Node.js 14 which doesn't support AbortController

eps1lon committed:
Lower coverage threshhold

Jest is not suited for this check since has no
knowledge about the build matrix.
Codecov should check this instead

MichaelDeBoey committed:
chore: remove `styfle/cancel-workflow-action` usage (#4)

Commit:
Port tests from waitFor DOM

Commit:
Unify advanceTimersByTime and flushPromises

Ports https://github.com/testing-library/dom-testing-library/pull/1229

eps1lon committed:
Setup Codesandbox CI (#5)

Commit:
format

MatanBobi commented:
@eps1lon Do we still want to move forwards with this effort?

siriwatknp commented on src/waitFor.ts:
```diff
@@ -0,0 +1,157 @@
+
+export default function waitFor(
+  callback: () => void | Promise<void>,
+  options: WaitForOptions,
```
```suggestion
  options?: WaitForOptions,
```
Should `options` be optional? Otherwise, I get a TS error if 2nd arg is not provided.
